### PR TITLE
feat(precompile): add biguint arithmetic precompiles

### PR DESCRIPTION
**Pull request by 0xkanekiken**
`Mul` precompile cycle:

Screenshot 2024-03-15 at 8 17 34 PM

`Div` precompile cycle:

Screenshot 2024-03-16 at 3 56 04 AM

When no dedicated tables were used:

Screenshot 2024-03-15 at 3 49 46 PM

### CODE REVIEW

File: core/src/syscall/precompiles/uint256/uint256_mul.rs [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+fn biguint_from_words(words: &[u32]) -> BigUint {
+    let bytes = words
+        .iter()
+        .flat_map(|n| n.to_le_bytes())
+        .collect::<Vec<_>>();
+    BigUint::from_bytes_le(bytes.as_slice())
+}
+
+// Convert a BigUint into a vector of u32 words
+fn biguint_to_words(value: &BigUint) -> [u32; NUM_WORDS_IN_UINT256] {
```
Comment (ctian1) [inline core/src/syscall/precompiles/uint256/uint256_mul.rs:307]: we can use `bytes_to_words_le`, its basically the same as this function. same with `words_to_bytes_le`

File: zkvm/precompiles/src/lib.rs [hunk 2]
```diff
@@ -14,6 +15,8 @@ extern "C" {
     pub fn syscall_secp256k1_double(p: *mut u32);
     pub fn syscall_secp256k1_decompress(point: &mut [u8; 64], is_odd: bool);
     pub fn syscall_keccak_permute(state: *mut u64);
+    pub fn syscall_uint256_mul(x: *mut u32, y: *mut u32);
+    pub fn syscall_keccak256_compress(state: *mut u32, message: *const u32);
```
Comment (ctian1) [inline zkvm/precompiles/src/lib.rs:19]: delete this

File: zkvm/precompiles/src/uint256_div.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
+use crate::io;
+use crate::syscall_uint256_mul;
+use crate::unconstrained;
+use num::{BigUint, Integer};
+
+/// Uint256 division operation.
+pub fn uint256_div(x: &[u8; 32], y: &mut [u8; 32]) -> [u8; 32] {
```
Comment (ctian1) [inline zkvm/precompiles/src/uint256_div.rs:7]: Please add a descriptive comment, and note that y is modified to be x which is a bit unexpected

File: zkvm/precompiles/src/uint256_div.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                syscall_uint256_mul(y.as_mut_ptr() as *mut u32, quotient_bytes.as_mut_ptr() as *mut u32);
+            }
+
+            let quotient_times_divisor = BigUint::from_bytes_le(y);
+            assert_eq!(quotient_times_divisor, dividend - remainder);
+
+            quotient_bytes
+        } else {
+            let result_biguint = BigUint::from_bytes_le(x) / BigUint::from_bytes_le(y);
+            result_biguint.to_bytes_le().try_into().unwrap_or([0; 32])
```
Comment (ctian1) [inline zkvm/precompiles/src/uint256_div.rs:41]: I think you need to do .resize(32, 0u8) before try_into

**Review by ctian1 — CHANGES_REQUESTED** [review]

File: core/src/syscall/precompiles/uint256/mod.rs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+use crate::utils::ec::field::{FieldParameters, NumLimbs};
+use num::{BigUint, One};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+pub struct U256Field;
+
+impl FieldParameters for U256Field {
+    const MODULUS: &'static [u8] = &[
```
[the remark this answers is not in the document]
Comment (0xkanekiken) [reply core/src/syscall/precompiles/uint256/mod.rs:15]: yeah, agreed! I fixed it in the latest commit.

File: core/src/syscall/precompiles/uint256/mod.rs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    ];
+
+    const WITNESS_OFFSET: usize = 1usize << 13;
```
[the remark this answers is not in the document]
Comment (0xkanekiken) [reply core/src/syscall/precompiles/uint256/mod.rs:21]: I've borrowed it from secp256k1 and bn254.

File: zkvm/precompiles/src/uint256_div.rs (hunk 3 above)
Comment (0xkanekiken) [reply zkvm/precompiles/src/uint256_div.rs]: I've taken a mutable reference of x instead of y in the latest commit

**Review by ctian1 — APPROVED** [review]

File: core/src/syscall/precompiles/uint256/uint256_mul.rs [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+            local.is_real,
+        );
+
+        builder.constraint_memory_access_slice(
+            local.shard,
+            local.clk + AB::F::from_canonical_u32(1), // We read p at +1 since p, q could be the same.
+            local.x_ptr,
+            &local.x_memory,
+            local.is_real,
+        );
```
[the remark this answers is not in the document]
Comment (0xkanekiken) [reply core/src/syscall/precompiles/uint256/uint256_mul.rs:277]: done!

File: core/src/syscall/precompiles/uint256/uint256_mul.rs [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+        // Generate the trace rows & corresponding records for each chunk of events concurrently.
+        let rows_and_records = input
+            .uint256_mul_events
+            .par_chunks(chunk_size)
+            .map(|events| {
+                let mut records = ExecutionRecord::default();
+                let mut new_byte_lookup_events = Vec::new();
+
+                let rows = events
+                    .iter()
```
[the remark this answers is not in the document]
Comment (0xkanekiken) [reply core/src/syscall/precompiles/uint256/uint256_mul.rs:98]: We won't be able to as then we won't be able to borrow new_byte_lookup_events as a mutable ref inside a closure

File: tests/uint256_div/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+use std::panic;
+
+use sp1_zkvm::precompiles::uint256_div::uint256_div;
+
+#[sp1_derive::cycle_tracker]
+fn main() {
+    // Random test.
+    for _ in 0..100 {
+        // generate random dividend and divisor.
+        let mut rng = rand::thread_rng();
```
[the remark this answers is not in the document]
Comment (0xkanekiken) [reply tests/uint256_div/src/main.rs:16]: I printed the values of x and y and got the following:

```
stdout: WARNING: Using insecure random number generator
stdout: Multiplying x: [66, 230, 172, 162, 225, 146, 169, 79, 163, 80, 123, 92, 163, 98, 231, 193, 177, 72, 71, 22, 44, 155, 220, 109, 80, 165, 135, 9, 223, 107, 220, 65] by y: [243, 130, 212, 52, 140, 233, 143, 20, 28, 117, 224, 147, 24, 75, 37, 212, 109, 252, 202, 188, 69, 242, 111, 220, 74, 88, 221, 93, 224, 45, 194, 40]
stdout: Multiplying x: [55, 24, 130, 136, 246, 65, 202, 90, 246, 203, 148, 79, 134, 235, 86, 93, 176, 242, 57, 83, 63, 184, 77, 26, 37, 51, 254, 71, 217, 226, 249, 87] by y: [99, 249, 175, 218, 108, 104, 149, 120, 118, 197, 168, 115, 34, 33, 67, 233, 82, 58, 74, 164, 188, 56, 6, 98, 174, 85, 33, 3, 197, 57, 212, 161]
stdout: Multiplying x: [138, 128, 246, 87, 124, 63, 232, 133, 58, 251, 233, 186, 58, 4, 62, 101, 133, 251, 109, 80, 58, 247, 197, 215, 146, 211, 127, 203, 129, 22, 185, 145] by y: [73, 14, 235, 216, 93, 241, 58, 204, 80, 196, 57, 180, 29, 58, 230, 57, 210, 21, 225, 170, 20, 18, 151, 149, 89, 11, 189, 26, 52, 79, 223, 8]
stdout: Multiplying x: [184, 168, 199, 240, 52, 194, 179, 250, 92, 76, 130, 7, 9, 35, 222, 108, 3, 148, 194, 218, 163, 183, 209, 175, 210, 47, 118, 255, 227, 118, 88, 255] by y: [229, 52, 236, 177, 218, 80, 119, 151, 32, 171, 39, 178, 150, 107, 226, 205, 183, 36, 85, 150, 177, 164, 218, 61, 12, 187, 114, 4, 83, 100, 60, 215]
stdout: Multiplying x: [240, 53, 138, 160, 11, 117, 51, 162, 61, 54, 94, 205, 210, 82, 251, 19, 120, 72, 213, 198, 157, 26, 215, 122, 59, 59, 134, 46, 231, 20, 224, 56] by y: [245, 18, 87, 114, 118, 241, 211, 186, 226, 86, 111, 208, 105, 18, 54, 59, 13, 172, 67, 181, 222, 0, 182, 97, 106, 25, 110, 4, 204, 157, 204, 66]
```

It gave a warning about using an insecure random number generator. Please let me know how to deal with this.

**Review by jtguibas — APPROVED** [review]

**Pull request merged**